### ISSUE TITLE
Consider making the AppImage self-updatable from the menu

### ISSUE THREAD

**Issue by probonopd** [FreeCAD/FreeCAD-AppImage#27]
Now that FreeCAD is promoting https://www.freecadweb.org/wiki/Appimage#Auto_Updating, why not make the AppImage self-updatable from the menu? Possibly using https://github.com/antony-jr/AppImageUpdaterBridge

cc @antony-jr

**Comment by luzpaz** [comment]
Discussion thread: https://forum.freecadweb.org/viewtopic.php?f=8&t=44324

**Comment by antony-jr** [comment]
Look into https://github.com/TheAssassin/python-appimageupdate but it comes with a lot of issues because the library uses **cffi**, See https://github.com/Pext/Pext/issues/95

I once used Qt plugins with PyQt with no problem, But it might take some time to make documentation on that.

**Comment by luzpaz** [comment]
@antony-jr Nice! Is it compatible with pyside2. I don't think FreeCAD is using PyQt5, I may be wrong.

**Comment by antony-jr** [comment]
> @antony-jr Nice! Is it compatible with pyside2? I don't think FreeCAD is using PyQt5, I may be wrong.

It should if it has **QPluginLoader** implemented.

**Comment by triplus** [comment]
Thanks for the detailed description. We discuss such things occasionally on the forum (for all platform binaries). Realistically likely we won't be implementing such integrated updating support in foreseeable future. The strategy of using an external updating application works good. Once you embed an updater tools you have to start taking care of potential update related issues. And overall keep everything in good shape. That is not desirable on our side ATM.

@sgrogan 

Could you add a note mentioning:

https://github.com/antony-jr/AppImageUpdaterBridge

Directly to releases page?

In short i feel that people using AppImages can handle a drag and drop operation just fine, for updating. And that already is an elegant and easy solution without introducing any additional maintenance burden and potential issues.

**Comment by triplus** [comment]
The report can stay opened for possible future re-evaluation. If somebody else would like to go ahead with this. I am OK with that.

**Comment by luzpaz** [comment]
hmm... I don't think this is such a heavy burden. If it breaks then we comment it out. I say we try it and see. This is an elegant solution in itself. I vote to test it out

**Comment by triplus** [comment]
Sure, go ahead. I don't object others trying it out.

**Comment by triplus** [comment]
If at some point we decide to move forward, what we can do is to first pack the library on Conda and for the AppImage to pick it up.

P.S. There is a plethora of Python developers using Linux on the FreeCAD forum and after they will gain access to such library from the FreeCAD Python console ...

**Comment by looooo** [comment]
I tried with PySide2 but this wasn't successful:

```
>>> from PySide.QtCore import QPluginLoader
>>> path = os.path.join(os.environ["CONDA_PREFIX"], "lib", "libAppImageUpdaterBridge.so")
>>> from PySide.QtCore import QPluginLoader
>>> loader = QPluginLoader()
>>> loader.setFileName(path)
>>> loader.load()
False
>>> loader.errorString()
"Failed to extract plugin meta data from '~/miniconda3/envs/freecad/lib/libAppImageUpdaterBridge.so'"
```

any ideas?

**Comment by antony-jr** [comment]
@looooo Did you build the plugin with Qt version lesser than or equal to the Qt version which loads the plugin?

**Comment by luzpaz** [comment]
Nice, we also have a [conda recipe](https://forum.freecadweb.org/viewtopic.php?f=8&t=44324&p=379151#p379151) for the appimage-updater-bridge thanks to loooo

**Comment by antony-jr** [comment]
@looooo I guess you built the library with the stable version **v1.1.6**, if that is true then the problem is that the plugin feature is in alpha, it only exists in the ```plugin``` branch. You have to use that branch to get a valid Qt Plugin.

**Comment by antony-jr** [comment]
Please see https://forum.freecadweb.org/viewtopic.php?f=8&t=44324&start=10#p381756

**Comment by looooo** [comment]
thanks a lot for your work. The package is updated and works:
```
>>> from PySide.QtCore import QPluginLoader
>>> path = os.path.join(os.environ["CONDA_PREFIX"], "lib", "libAppImageUpdaterBridge.so")
>>> from PySide.QtCore import QPluginLoader
>>> loader = QPluginLoader()
>>> loader.setFileName(path)
>>> loader.load()
True
```

**Comment by probonopd** [comment]
Wow, great progress. What do people think, can this be used?

**Comment by antony-jr** [comment]
> Wow, great progress. What do people think, can this be used?

The library can be used in Python with ease but we are still testing it in FreeCAD console which seems to have some issues. 

See https://antony-jr.github.io/AppImageUpdaterBridge/docs/PyQt5PluginExample.html

**Comment by antony-jr** [comment]
The issue seems to be specific with FreeCAD's console, Please see https://forum.freecadweb.org/viewtopic.php?f=8&t=44324&p=401553#p401553

**Comment by antony-jr** [comment]
@looooo I guess we should start using the GUI directly, considering https://forum.freecadweb.org/viewtopic.php?f=8&t=44324&p=401553#p401553

I'm almost done with the changes.

**Comment by antony-jr** [comment]
@looooo @luzpaz Please see https://forum.freecadweb.org/viewtopic.php?f=8&t=44324&p=432932#p432932 for the error on updating when AppImage Launcher is installed.

**Comment by antony-jr** [comment]
@probonopd Starting from FC 0.19.21514 (Development Build) there now exists an AppImage section found via **Edit → Preferences → AppImage.** 

![self-update](https://user-images.githubusercontent.com/33930174/94679848-673f7600-033e-11eb-88e9-6002357cd97d.png)

Special thanks to @looooo

**Comment by probonopd** [comment]
Congratulations and thank you to everyone involved.

**Comment by luzpaz** [comment]
Thanks to all invovled
